packages/components/src/internal/OverlayTrigger.tsx

```diff
@@ -0,0 +1,137 @@
+import React, {
+    cloneElement,
+    Children,
+    FC,
+    useRef,
+    ReactElement,
+    useState,
+    useCallback,
+    createElement,
+    MutableRefObject,
+} from 'react';
+import { createPortal } from 'react-dom';
+
+import classNames from 'classnames';
+
+import { usePortalRef } from './hooks';
+
+interface OverlayTriggerState<T extends Element = HTMLDivElement> {
+    onMouseEnter: () => void;
+    onMouseLeave: () => void;
+    onClick: () => void;
+    portalEl: HTMLElement;
+    show: boolean;
+    targetRef: MutableRefObject<T>;
+}
+
+/**
+ * useOverlayTriggerState is a useful hook for when you want to render an overlay, but you can't use OverlayTrigger
+ * because you can't wrap your component in an arbitrary div.
+ * @param id the id to use for the portal
+ * @param hoverEventsEnabled whether onMouseEnter/Leave will trigger the show state
+ * @param clickEventEnabled whether onClick with trigger the showState
+ * @param delay the amount, in ms, to delay before showing the overlay
+ */
+export function useOverlayTriggerState<T extends Element = HTMLDivElement>(
+    id: string,
+    hoverEventsEnabled: boolean,
+    clickEventEnabled: boolean,
+    delay: number = undefined
+): OverlayTriggerState<T> {
+    const targetRef = useRef<T>(null);
+    const portalEl = usePortalRef('overlay-trigger-portal-' + id);
+    const [show, setShow] = useState<boolean>(false);
+    const [_, setTimeoutId] = useState<number>(undefined);
+    const onMouseEnter = useCallback(() => {
+        if (!hoverEventsEnabled) return;
+
+        if (delay) {
+            setTimeoutId(window.setTimeout(() => setShow(true), delay));
+            return;
+        }
+
+        setShow(true);
+    }, [delay, hoverEventsEnabled]);
+    const onMouseLeave = useCallback(() => {
+        if (!hoverEventsEnabled) return;
+
+        setTimeoutId(currentId => {
+            if (currentId) window.clearTimeout(currentId);
+            return undefined;
+        });
+
+        setShow(false);
+    }, [hoverEventsEnabled]);
+    const onClick = useCallback(() => {
+        if (!clickEventEnabled) return;
+        setShow(_show => !_show);
+    }, [clickEventEnabled]);
+
+    return {
+        onClick,
+        onMouseEnter,
+        onMouseLeave,
+        portalEl,
+        show,
+        targetRef,
+    };
+}
+
+export interface OverlayComponent<O extends Element = HTMLDivElement> {
+    targetRef?: MutableRefObject<O>;
+}
+
+export type TriggerType = 'click' | 'hover';
```

Comment:
Consider exporting `TriggerType` in `index.ts`. I've found that these combination types don't export well across packages so it's easier if users of the types can reference the type directly.